### PR TITLE
port world time and light to lua and syncing world time with real time

### PR DESCRIPTION
**Pull request by Sundance**
When tfs is started it counts worldTime from 14:27 when it should be synced with realTime (hh:00 is always 00:00 in-game, hh:30 is always 12:00 in-game, etc)
Porting it to lua will also give the host more control such as modifying the time or light however they wish.
A heads up: The default client has a fade function that makes light changes happen really slowly.
So if anyone wants to create an area that's always dark otclient or otu-client is the way to go

### CODE REVIEW

File: data/globalevents/scripts/sendworldlight.lua [hunk 1]
```diff
@@ -0,0 +1,11 @@
...
+function onThink(interval)
+	local newIntensity,newColor = getWorldLight()
+	if newIntensity ~= intensity or newColor ~= color then
+		for k,v in pairs(Game.getPlayers()) do
+			v:sendWorldLight(intensity, color)
+		end
+		intensity,color = getWorldLight()
+	end
+	return true
+end
```
Comment (EPuncker) [inline data/globalevents/scripts/sendworldlight.lua]: add new line at end of file

Comment (Sundance) [reply data/globalevents/scripts/sendworldlight.lua]: Done

File: data/creaturescripts/scripts/login.lua [hunk 2]
```diff
@@ -1,4 +1,6 @@
 function onLogin(player)
+	local intensity,color = getWorldLight()
```
Comment (DSpeichert) [inline data/creaturescripts/scripts/login.lua]: `intensity, color`

File: data/global.lua [hunk 3]
```diff
@@ -23,15 +23,41 @@ function getDistanceBetween(firstPosition, secondPosition)
...
+	
+	worldMinutes = worldMinutes%60
 
-	local minutes = worldTime % 60
-	if minutes < 10 then
-		minutes = '0' .. minutes
+	if worldHours < 10 then hoursPrefix = "0" end
+	if worldMinutes < 10 then minutesPrefix = "0" end
+
+	return hoursPrefix..worldHours..":"..minutesPrefix..worldMinutes
```
Comment (DSpeichert) [inline data/global.lua]: `string.format("%02d:%02d", worldHours, worldMinutes)` ?
Not sure if we have this function available.

File: data/global.lua [hunk 4]
```diff
@@ -23,15 +23,41 @@ function getDistanceBetween(firstPosition, secondPosition)
...
+end
+
+do
+	local noon = 720
+	local changePerMinute = 0.29
+	function getWorldLight()
+		local intensity = nil
+		if getWorldTime() > noon then
+			intensity = math.floor(noon - (getWorldTime() - noon)*changePerMinute+40)
```
Comment (DSpeichert) [inline data/global.lua]: inconsistent formatting (spaces)

File: data/global.lua [hunk 5]
```diff
@@ -23,15 +23,41 @@ function getDistanceBetween(firstPosition, secondPosition)
...
+		else
+			intensity = math.floor(getWorldTime()*changePerMinute+40)
```
Comment (DSpeichert) [inline data/global.lua]: inconsistent formatting (spaces)

File: data/globalevents/scripts/sendworldlight.lua [hunk 6]
```diff
@@ -0,0 +1,11 @@
+local intensity,color = getWorldLight()
+function onThink(interval)
+	local newIntensity,newColor = getWorldLight()
```
Comment (DSpeichert) [inline data/globalevents/scripts/sendworldlight.lua]: inconsistent formatting (spaces)

File: data/globalevents/scripts/sendworldlight.lua [hunk 7]
```diff
@@ -0,0 +1,11 @@
...
+	if newIntensity ~= intensity or newColor ~= color then
+		for k,v in pairs(Game.getPlayers()) do
+			v:sendWorldLight(intensity, color)
+		end
+		intensity,color = getWorldLight()
```
Comment (DSpeichert) [inline data/globalevents/scripts/sendworldlight.lua]: inconsistent formatting (spaces)

**Comment by MillhioreBT** [comment]
I receive this debug error once I implement `port world time and light to lua and syncing world time with real time`
image

How can I repeat it?
* When the server expels the player for excess packets, when he tries to connect again, his character is left with that error until he is eliminated from the world.

This problem is immediately related to these modifications because you try the following:
image

Connect the 80 ~ 90% of the time generates exactly the same debug error.

**Comment by Sundance** [comment]
Closing because I'm not content with the approach taken.
There was no need to move everything to lua. The only thing that should be an option in lua imo is customizing worldlight.
There was also an issue with the approach. lightinfo has to be sent in ProtocolGame::sendAddCreature or the player's client will crash if trying to login to an already logged in character (my approach was only sending lightinfo on login)

File: src/game.h [hunk 8]
```diff
@@ -542,20 +528,10 @@ class Game
 
 		ModalWindow offlineTrainingWindow { std::numeric_limits<uint32_t>::max(), "Choose a Skill", "Please choose a skill:" };
 
-		static constexpr int32_t LIGHT_LEVEL_DAY = 250;
-		static constexpr int32_t LIGHT_LEVEL_NIGHT = 40;
-		static constexpr int32_t SUNSET = 1305;
-		static constexpr int32_t SUNRISE = 430;
```
Comment (Leo32onGIT) [inline src/game.h]: And these.

**Pull request closed without merging**